include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -3776,23 +3765,41 @@ namespace internal
       std::vector<Point<dim>> points;
       points.resize(sp_indices.size());
 
+      const auto locally_onwed_sp = dof_handler_sp.locally_owned_dofs();
+      std::vector<unsigned int> indices_state(locally_onwed_sp.n_elements(),
+                                              numbers::invalid_unsigned_int);
+
+      for (unsigned int i = 0; i < sp_indices.size(); ++i)
+        indices_state[sp_indices[i]] = i;
+
       const auto &  fe_sp = dof_handler_sp.get_fe();
       FEValues<dim> fe_values(mapping,
                               fe_sp,
                               Quadrature<dim>(fe_sp.get_unit_support_points()),
                               update_quadrature_points);
 
-      for (unsigned int i = 0; i < sp_indices.size(); ++i)
+      std::vector<types::global_dof_index> dof_indices(fe_sp.n_dofs_per_cell());
+
+      for (const auto &cell : dof_handler_sp.active_cell_iterators() |
+                                IteratorFilters::LocallyOwnedCell())
         {
-          typename DoFHandler<dim>::active_cell_iterator cell_sp = {
-            &dof_handler_sp.get_triangulation(),
-            std::get<0>(sp_cells[i]),
-            std::get<1>(sp_cells[i]),
-            &dof_handler_sp};
-          fe_values.reinit(cell_sp);
-          points[i] = fe_values.quadrature_point(std::get<2>(sp_cells[i]));
-        }
+          fe_values.reinit(cell);
+          cell->get_dof_indices(dof_indices);
+
+          for (const unsigned int q : fe_values.quadrature_point_indices())
+            if (locally_onwed_sp.is_element(dof_indices[q]))
+              {
+                const auto index =
+                  locally_onwed_sp.index_within_set(dof_indices[q]);
 
+                if (indices_state[index] != numbers::invalid_unsigned_int)
+                  {
+                    points[indices_state[index]] =
+                      fe_values.quadrature_point(q);
+                    indices_state[index] = numbers::invalid_unsigned_int;
+                  }
+              }
+        }
```

Comment:
If we do it like this, we have to check if the DoFs are constrained again. If a DoF is constrained, the vector of Points is too small, leading to a segfault?

Reply:
I don't think so. `indices_state[index] == numbers::invalid_unsigned_int` for the constrained points, isn't it?

Reply:
For me, all tests are passing.

Reply:
True, I confused the definition of `locally_onwed_sp` with `indices_state`

Reply:
The reason for this modification is that setting up `FEValues` for each point is quite expensive. Let's consider high-order FEM: we would set up `x (k)^d` too often.